### PR TITLE
Autogenerate go-md2man(able) md file for kubectl

### PR DESCRIPTION
**Pull request by eparis**
This does away with the giant dump from cobra for kubectl and instead
generates md files which can then be post processed into man pages.
I'll work on automating that as well instead of needing the outside
tool.

I may take these ideas to the cobra project to see if they want to try
to do it there instead, but this gets us something much more manageable
and less overwhelming today.

### CODE REVIEW

**Comment by brendandburns** [comment]
Thanks for doing this, I think it will be a big improvement.
@jlowdermilk for his thoughts.

Some feedback:
   * The MD files are pretty ugly (esp. the % noise at the top)
   * There's no way to navigate from kubectl.md to the sub-commands, which makes navigating the web docs pretty annoying
   * If we really are going to delete kubectl.md, then we need to update docs/cli.md to point to the right place(s)

**Comment by brendandburns** [comment]
Also, needs re-base.

**Comment by jlowdermilk** [comment]
Initial thoughts:
- Thanks for doing this, having autogenerated man pages is awesome.
- While a bit clunky, a single kubectl.md doc is useful in that it's easy to verify docs are up to date, and as serves as a base file that can be parsed to render docs elsewhere. E.g., we want to use it in GKE to generate web docs for the kubectl release that we ship with the cloud sdk (usually lags k8s head by 1-2 weeks).

Can we keep kubectl.md, at least until we have a way to automate verifying that the manable docs are up to date?

**Comment by eparis** [comment]
This patch does all of the same verification...

**Comment by jlowdermilk** [comment]
So it does; not sure how I missed that. lgtm

**Comment by eparis** [comment]
I totally understand @brendandburns comments about the preample cruft.  The .md files could be a whole lot more useful. I admit, my finished goal was the man pages, and the .md files are just a necessary intermediate state. Maybe I can work with the upstream projects to get a better set of .md files and go directly to man pages...

**Comment by brendandburns** [comment]
@eparis I would be happy to merge this as is, if we also leave kubectl.md around.  (and the presubmit validation too)

So I guess we can take an iterative approach and do both kubectl.md and the man page generation while we work on cleaning up the generated MD, or we can try to do it all in the context of this PR.

Your call.

Thanks
--brendan

**Comment by brendandburns** [comment]
(oh, and it still needs a rebase ;)

**Comment by eparis** [comment]
I'll put the giant file dump back, and keep iterating on the autogenerated .md's in another PR.  I'll let you know when it is ready.

**Comment by brendandburns** [comment]
@eparis cool, thanks!
--brendan

**Comment by eparis** [comment]
check this out, better .md web docs and autogenerated man pages....

notice the web docs show the parent verb's command line options, but the man page doesn't.  probably should show them in the man page as well...

**Comment by jlowdermilk** [comment]
Will look over again this afternoon, I expect we'll be good to merge.

File: cmd/genman/gen_kubectl_man.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+			fmt.Fprintf(out, "**%s-%s(1)**, ", dname, c.Name())
+			genMarkdown(c, name, docsDir)
+		}
+		fmt.Fprintf(out, "\n")
+	}
+
+	out.WriteString(`
+# HISTORY
+January 2015, Originally compiled by Eric Paris (eparis at redhat dot com) based on the kubernetes source material, but hopefully they have been automatically generated since!
+`)
```
Comment (jlowdermilk) [inline cmd/genman/gen_kubectl_man.go]: Optional/TODO: write the current date here, e.g. "last compiled: MMM DD YYYY", so it gets updated everytime we run `hack/run-gendocs.sh`.

**Comment by jlowdermilk** [comment]
lgtm. I'm happy to merge as is if @brendandburns has no reservations. 

Sidenote: the splitting of owned vs inherited options in the markdown is excellent; the approach might be useful for #4142.

**Comment by jlowdermilk** [comment]
Needs rebase.

**Comment by brendandburns** [comment]
Can you add LICENSE files for mangen and blackfriday.  Then LGTM.

--brendan

**Comment by eparis** [comment]
https://github.com/cpuguy83/go-md2man/pull/9
https://github.com/russross/blackfriday/pull/145

The licenses are already in the README files.  And I don't want to make changes in Godeps by hand, won't they get overwritten next time the Godeps are updated?

**Comment by eparis** [comment]
I also don't understand the LICENSE comment/request at all, now that I think about it.  The LICENSE lives in the root of the go-md2man/ directory, but Godeps is only pulling in go-md2man/mangen

If you look there are a bunch of projects under Godeps where the LICENSE of the Godep isn't in our repo, which might actually be a problem...

**Comment by saad-ali** [comment]
@eparis Please rebase
@brendandburns Once you give the thumbs up, I'll merge this

**Comment by brendandburns** [comment]
Lgtm, given that the license text is in the readme.

**Comment by eparis** [comment]
rebased

**Comment by saad-ali** [comment]
@eparis  Both Shippable and travis builds are failing with ```The command "PATH=$HOME/gopath/bin:./third_party/etcd:$PATH ./hack/verify-gendocs.sh" exited with 1.
``` Could you take a look

**Comment by jlowdermilk** [comment]
Looks like someone changed whitespace in the help text for labels. Rebase and rerun gendocs should fix. I will greedily merge this on green, been wanting to get it in for some time.

**Comment by eparis** [comment]
I rebased, but see no changes when i regen the docs with this new code...

**Comment by eparis** [comment]
hmmmmm, interesting, my local machine is mis-generating the man1 kubectl-label.1 file, but travis is generating it correctly!

**Comment by jlowdermilk** [comment]
Yeah, this appears to be a problem with the help text itself. The label commands examples include text enclosed with `<...>` which mangen appears to not handle well. If you don't mind, the easiest fix is probably to just edit [label.go](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/cmd/label.go) and replace the `<...>` with `//`, which matches the rest of the commands anyway.

**Comment by eparis** [comment]
ok, hopefully I've got it fixed now.  I'm not 100% sure what I did to fix it.  I updated from golang 1.3 to 1.4.  (which was unlikely the problem)  I also removed _output and Godeps/_workspace/pkg.  Likely it was a problem with and old .a file in my local builds.  Hopefully we'll go green here in just a second.

**Comment by eparis** [comment]
ok, this failure doesn't look like my fault!

**Comment by eparis** [comment]
@jlowdermilk ok, now travis passed and shippable doesn't...   "details" just shows me their homepage.  not sure what to do

**Comment by jlowdermilk** [comment]
Yeah, you have to log in with github auth before it shows you build results. Anyway, that shippable failure looks like a flake.

**Comment by bgrant0607** [comment]
It's been 5 days. Re-running Travis before merging.

**Comment by nikhiljindal** [comment]
Travis passed for 1.4 but failed for 1.3.
Have triggered it again to see if it was a fluke.

**Comment by jlowdermilk** [comment]
Unfortunately, this needs yet another rebase.

**Comment by jlowdermilk** [comment]
Since this keeps getting blocked due to other changes with merge conflicts, I'd say feel free to self-merge whenever you have time to rebase and get a clean travis run.

**Comment by brendandburns** [comment]
I'm self-merging this.

**Pull request merged**